src/scores/contingency.py

```diff
@@ -0,0 +1,258 @@
+
+A simple example is where Python boolean True and False are used to indicate a binary match.
+An alternative could be to use the integers 0 or 1.
+
+A more complex example is that of categories. Category designations could equally be a series of 
+numbers (e.g. 0, 1, 2, 3); "buckets" or ranges (e.g. 0-.9, 1-1.9, 2-2.9), or labels ("rain", "no rain")
+or yet more complex examples and data types. Some practical decisions have to be made to provide a
+useful implementation which is computationally efficient while being sufficiently flexible to 
+support the expected requirements. The developers acknowledge some of these choices may not
+match all user preferences, and would happily receive any feedback about alternative approaches.
+
+The implementation in this module uses "True" and "False" for the binary table, and uses integers
+for categorical data. Users wishing to match ranges or named categories should maintain a separate
+data structure mapping between the integer values and the semantics of the category. For convenience,
+all tables have a 'translation' dictionary which can be used to reference and dereference the categories.
+
+Users can supply their own matching operators to the top-level module functions so long as
+they can translate from a score (or proximity function) to either a BinaryTable or a 
+CategoryTable.
+'''
+
+import operator
+import xarray
+
+DEFAULT_PRECISION = 8
+
+def make_binary_table(proximity, match_operator):
+	table = match_operator.make_table(proximity)
+	return table
+
+def make_category_table(proximity, category_operator):
+	table = category_operator.make_table(proximity)
+	return table
+
+class BinaryContingencyTable():
+	'''
+	At each location, the value will either be:
+	 - A true positive    (0)
+	 - A false positive   (1)
+	 - A true negative    (2)
+	 - A false negative   (3)
+
+	It will be common to want to operate on masks of these values,
+	such as:
+	 - Plotting these attributes on a map
+	 - Calculating the total number of these attributes
+	 - Calculating various ratios of these attributes, potentially
+	   masked by geographical area (e.g. accuracy in a region)
+
+	As such, the per-pixel information is useful as well as the overall
+	ratios involved.
+	'''
+
+	def __init__(self, forecast_events, observed_events):
+
+		# type checking goes here
+		self.forecast_events = forecast_events
+		self.observed_events = observed_events
+
+		self.tp = (self.forecast_events == 1) & (self.observed_events == 1)  # true positives
+		self.tn = (self.forecast_events == 0) & (self.observed_events == 0)  # true negatives
+		self.fp = (self.forecast_events == 1) & (self.observed_events == 0)  # false positives
+		self.fn = (self.forecast_events == 0) & (self.observed_events == 1)  # false negatives
+
+		# Variables for count-based metrics, calculated on first access
+		self.tp_count = self.tp.sum().values.item()     # Count of true positives
```

Comment:
```suggestion
		self.tp_count = self.tp.sum()    # Count of true positives
```
We wan't to return an xarray object rather than an int so that dask works nicely and so that we can return counts with dims preserved (e.g., lead day).

Reply:
And within the `sum()` you can specify the dims to reduce, which could be specified when the object is initialised.

Reply:
Do you think people would want a flag to get a float instead? Or just leave that to the user?

Reply:
I think leave it to the use to decide. That is how most the other metrics in scores work

Reply:
Done.